quarkchain/cluster/jsonrpc.py

```diff
@@ -1465,6 +1469,76 @@ def shutdown(self):
 
             await asyncio.sleep(0.5)
 
+    async def fetch_logs(self, sub_id, full_shard_id, params, websocket):
+        decoder = address_decoder
+        start_block = params.get("fromBlock", "latest")
+        end_block = params.get("toBlock", "latest")
+        # TODO: not supported yet for "earliest" or "pending" block
+        if (isinstance(start_block, str) and start_block != "latest") or (
+            isinstance(end_block, str) and end_block != "latest"
+        ):
+            return None
+        # parse addresses / topics
+        addresses, topics = [], []
+        if "address" in params:
+            if isinstance(params["address"], str):
+                addresses = [Address.deserialize(decoder(params["address"]))]
+            elif isinstance(params["address"], list):
+                addresses = [Address.deserialize(decoder(a)) for a in params["address"]]
+        if full_shard_id is not None:
+            addresses = [Address(a.recipient, full_shard_id) for a in addresses]
+        if "topics" in params:
+            for topic_item in params["topics"]:
+                if isinstance(topic_item, str):
+                    topics.append([data_decoder(topic_item)])
+                elif isinstance(topic_item, list):
+                    topics.append([data_decoder(tp) for tp in topic_item])
+
+        branch = Branch(full_shard_id)
+        shard = self.slave.shards.get(branch, None)
+        header = shard.state.header_tip
+        if start_block == "latest":
```

Comment:
as mentioned before, should remove `start_block` and `end_block` related lines. instead, should use shard's tip to query latest logs